litestar/datastructures/headers.py

```diff
@@ -83,10 +86,13 @@ def from_scope(cls, scope: "HeaderScope") -> "Headers":
         Raises:
             ValueError: If the message does not have a ``headers`` key
         """
-        if (headers := scope.get("_headers")) is None:
-            headers = scope["_headers"] = cls(scope["headers"])  # type: ignore[typeddict-unknown-key]
+        if "state" not in scope:
+            return cls(scope["headers"])
```

Comment:
We have a test that passes a "http.response.start" message to `from_scope()` here:

https://github.com/litestar-org/litestar/blob/8b621ac13820840f2c48123b603340ec09ff1380/tests/unit/test_datastructures/test_headers.py#L77-L82

However, I cannot see anywhere in the library source where we actually do this - so I wonder if this is just some artifact that can be removed?

@provinzkraut, do you have any insight on this? The behavior that we test would have added a "_headers" key into the message in the old behavior which I'm not sure is what we want, and the new issue is that the asgi messages don't have a "state" key.

Reply:
I do. We did the `_`-prefixed thing in the top level scope before we introduces the `__litestar__` scope key. Iirc, when that change was made, only those instances where we explicitly set state (e.g. stuff we did in `set_litestar_scope_state`) were migrated.

Actually, I think it makes *more* sense to store our scope state within a top level `__litestar__` key than within `state`, as `state` is an optional key used by the [lifespan protocol](https://asgi.readthedocs.io/en/latest/specs/lifespan.html), to handle [lifespan state](https://asgi.readthedocs.io/en/latest/specs/lifespan.html#lifespan-state), which has some extra functionality attached to it, if supported by the server, so we're kinda interfering with that; It's more meant for things like our `State`.

Reply:
To clarify my question, was caching the headers in the `http.response.start` message a particular behavior we were testing for? We don't pass a `http.response.start`, or `websocket.accept` message to `Headers.from_scope()` in our source code anywhere, but we do it in this test.

> Actually, I think it makes more sense to store our scope state within a top level __litestar__ key than within state, as state is an optional key used by the [lifespan protocol](https://asgi.readthedocs.io/en/latest/specs/lifespan.html), to handle [lifespan state](https://asgi.readthedocs.io/en/latest/specs/lifespan.html#lifespan-state), which has some extra functionality attached to it, if supported by the server, so we're kinda interfering with that; It's more meant for things like our State.

That makes sense but it isn't related to this PR, so we can do it in a follow up.

Reply:
> To clarify my question, was caching the headers in the `http.response.start` message 

Yeah, no idea either. Doesn’t make any sense really, probably just a bugged test.

> That makes sense but it isn't related to this PR, so we can do it in a follow up.

Yes, let’s do that once this one gets merged :)

Reply:
This gives me pause on whether `ScopeState` is the right name for the caching object then. If we are conflating its usage, then surely we are conflating terms also.